### PR TITLE
3D plot GUI plugin

### LINKED ISSUE #231: 3D plot plugin
Gazebo-classic's [LinkPlot3DPlugin](https://github.com/osrf/gazebo/blob/gazebo11/plugins/LinkPlot3DPlugin.hh) traces the trajectory of a link in the rendering scene by publishing 3D markers.

Try `gazebo worlds/plot3d.world `:

plot3d

It would be nice to port it to Ignition. We already have markers, so it should be a matter of implementing a system plugin that publishes markers according to the current link poses.

### PR DESCRIPTION
**Pull request by chapulina**
# 🎉 New feature

Closes #231 

Alternative to #332

## Summary

Plots the path of entities in the 3D scene, parameters are customizable from SDF and / or the GUI.

~~WIP while I work on some final usability tweaks, and possibly a test.~~

Ready for review!

## Test it

```
ign gazebo -r plot_3d.sdf
```

This is what it currently looks like:

plot3d

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

File: examples/worlds/plot_3d.sdf
Comment (ahcorde) [inline examples/worlds/plot_3d.sdf:51]:
```suggestion
```

File: examples/worlds/plot_3d.sdf
Comment (ahcorde) [inline examples/worlds/plot_3d.sdf:75]:
```suggestion
```

File: include/ignition/gazebo/Util.hh
Comment (ahcorde) [inline include/ignition/gazebo/Util.hh:55]: name doesn't mach

File: src/gui/plugins/plot_3d/Plot3D_TEST.cc
Comment (ahcorde) [inline src/gui/plugins/plot_3d/Plot3D_TEST.cc:47]: it requires a initial value.

**Review by ahcorde — CHANGES_REQUESTED** [review]
I made a initial review, I'm going a tried in my machine

File: examples/worlds/plot_3d.sdf
Comment (chapulina) [reply examples/worlds/plot_3d.sdf:51]: Just curious, are you following any specific XML style for these suggestions? We haven't been enforcing any blank line policy so far.

File: src/gui/plugins/plot_3d/Plot3D_TEST.cc
Comment (chapulina) [reply src/gui/plugins/plot_3d/Plot3D_TEST.cc]: 3c50512

File: src/gui/plugins/plot_3d/Plot3D.hh
Comment (chapulina) [reply src/gui/plugins/plot_3d/Plot3D.hh]: See acd159d

**Comment by ahcorde** [comment]
Resize looks weird. X and R are bigger than the other parameters.

resize

File: src/gui/plugins/plot_3d/Plot3D_TEST.cc
Comment (ahcorde) [inline src/gui/plugins/plot_3d/Plot3D_TEST.cc:110]: EXPECT_EQ( Expectation, value to test );

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Review by azeey — COMMENTED** [review]
Works great! Do you by any chance know why the color of the line changes over time? For example, in your video, the line for the lower arm starts out a light green color and goes to dark green.

File: include/ignition/gazebo/Util.hh
Comment (chapulina) [reply include/ignition/gazebo/Util.hh:72]: I thought I had mentioned somewhere but now I see it's not here. I targeted Fortress so it's easier to test with another project, but my plan is to retarget it to Citadel after the PR has been reviewed and forward-port it from there :ok_hand:

File: include/ignition/gazebo/Util.hh
Comment (chapulina) [reply include/ignition/gazebo/Util.hh:73]: Ah good point, I was living in a world with unique entity names :woman_facepalming: 

Requiring an entity may work well for server plugins that are attached to entities, but GUI plugins don't really have a reference entity, so it would be cumbersome to figure out what entity we're `relative_to`. And I don't like the idea of requiring fully scoped names mainly because it makes it difficult to use the same GUI config with various worlds.

So, I have an alternative suggestion in a1cf2841ea2c57f3ef1e4bad464e76cfe829916b, which is to return all entities that match the scoped name. Then the caller can decide what to do with all of them. Let me know whether that works for your use case. I may have pushed the cumbersomeness to your side :sweat_smile:

File: src/Util.cc
Comment (chapulina) [reply src/Util.cc]: Nope, see new implementation

**Comment by chapulina** [comment]
Thanks for the reviews, I believe I've addressed all comments :heavy_check_mark:

File: include/ignition/gazebo/Util.hh
Comment (azeey) [reply include/ignition/gazebo/Util.hh:73]: It does indeed become very cumbersome for system writers if they have to sift through a list of entities to determine which entity they should pick. If the issue is figuring out what the world name should be, I think that information is available in `GuiRunner`, so it shouldn't be too difficult to pass it on to `GuiSystem`s. Then, we can have scoped names in GUI configs without needing the world name

Comment (chapulina) [reply include/ignition/gazebo/Util.hh:78]: Fair enough. Added `relativeTo` in 673388d6d058adf8c4dd4d4bc0d89d65c7c82053.

I kept the logic returning multiple entities just in case. Gazebo isn't enforcing unique names, so it isn't an impossibility. I can also imagine this being used to return multiple entities on purpose, for example:

```
model1
  the_link
    visual
model2
  the_link
    visual
```

If I want all visuals from all `the_link`s, I can get them with `entitiesFromScopedName("the_link::visual", ecm)`

**Review by caguero — APPROVED** [review]
Looks great!

File: include/ignition/gazebo/Util.hh
Comment (chapulina) [reply include/ignition/gazebo/Util.hh:78]: @azeey , I'll wait on your ok on this function before merging

File: include/ignition/gazebo/Util.hh
Comment (azeey) [inline include/ignition/gazebo/Util.hh:75]: Can you update the `\return` comment so it reflects that an `unordered_set` is returned.

File: include/ignition/gazebo/Util.hh
Comment (azeey) [reply include/ignition/gazebo/Util.hh:78]: Looks good. Thanks!

**Review by azeey — APPROVED** [review]
Thanks for iterating. There are a couple of codecheck issues, otherwise LGTM!

**Comment by chapulina** [comment]
> Do you by any chance know why the color of the line changes over time? 

Oh, btw, no, I don't. I think this may be an issue on the rendering end though. I'll see if I gather enough to ticket an issue.

**Comment by chapulina** [comment]
Retargeted from Fortress to Citadel after iterating:

1. Rebased from `main` to `ign-gazebo3`
2. Updated target branch and label
3. Made necessary changes in 362570b 

If anyone wants to use this feature with Fortress before the PR is forward-ported, checkout this commit: 8025da67c85e07730e26f8cad41a7608a58b0a8e.

I'll merge this into Citadel if CI is happy.

**Pull request merged**
